File: domain/src/test/java/mixter/domain/core/subscription/FakeFollowerRepository.java
package mixter.domain.core.subscription;

import mixter.domain.identity.UserId;
import org.assertj.core.util.Maps;
import org.assertj.core.util.Sets;

import java.util.Map;
import java.util.Set;

public class FakeFollowerRepository implements FollowerRepository {
    Map<UserId, Set<UserId>> followers = Maps.newHashMap();

    @Override
    public Set<UserId> getFollowers(UserId followee) {
        return this.followers.getOrDefault(followee, Sets.newHashSet());
    }

    @Override
    public void saveFollower(UserId followee, UserId follower) {
        Set<UserId> userFollowers = followers.getOrDefault(followee, Sets.newHashSet());
        userFollowers.add(follower);
        followers.put(followee, userFollowers);
    }

    @Override
    public void removeFollower(UserId followee, UserId follower) {
        Set<UserId> userFollowers = followers.getOrDefault(followee, Sets.newHashSet());
        userFollowers.remove(follower);
        followers.put(followee, userFollowers);
    }
}

File: domain/src/test/scala/mixter/domain/subscription/FakeFollowerRepository.scala
package mixter.domain.subscription

import mixter.domain.identity.UserId
import org.scalatest.{Suite, SuiteMixin}

private[subscription] trait FollowerRepositoryFixture extends SuiteMixin {
  this: Suite =>
  def withFollowerRepository(test: FakeFollowerRepository => Any): Any = {
    val sessionRepository = new FakeFollowerRepository()
    test(sessionRepository)
  }
}

private[subscription] class FakeFollowerRepository extends FollowerRepository {
  private var subscriptions = Map.empty[UserId, Set[UserId]].withDefaultValue(Set.empty)

  override def getFollowers(authorId: UserId): Set[UserId] = subscriptions(authorId)

  override def saveFollower(followee: UserId, follower: UserId): Unit =
    subscriptions += followee -> (subscriptions(followee) + follower)

  override def removeFollower(followee: UserId, follower: UserId): Unit =
    subscriptions += followee -> (subscriptions(followee) - follower)
}